### ISSUE TITLE
(Static) shovels crash with socket_closed_unexpectedly and are never reconnected.

### ISSUE THREAD

**Issue by BtbN** [rabbitmq/discussions#21]
<details>
 <summary>System and Crash Info</summary>

* RabbitMQ version: 3.8.1
* Erlang version: 22.1.8

* RabbitMQ server and client application log files: http://paste.ubuntu.com/p/ZhmnWsQmYN/

* RabbitMQ plugin information via `rabbitmq-plugins list`
```
Listing plugins with pattern ".*" ...
 Configured: E = explicitly enabled; e = implicitly enabled
 | Status: * = running on rabbit@Encode-1
 |/
[  ] rabbitmq_amqp1_0                  3.8.1
[  ] rabbitmq_auth_backend_cache       3.8.1
[  ] rabbitmq_auth_backend_http        3.8.1
[  ] rabbitmq_auth_backend_ldap        3.8.1
[  ] rabbitmq_auth_backend_oauth2      3.8.1
[  ] rabbitmq_auth_mechanism_ssl       3.8.1
[  ] rabbitmq_consistent_hash_exchange 3.8.1
[  ] rabbitmq_event_exchange           3.8.1
[  ] rabbitmq_federation               3.8.1
[  ] rabbitmq_federation_management    3.8.1
[  ] rabbitmq_jms_topic_exchange       3.8.1
[E*] rabbitmq_management               3.8.1
[e*] rabbitmq_management_agent         3.8.1
[  ] rabbitmq_mqtt                     3.8.1
[  ] rabbitmq_peer_discovery_aws       3.8.1
[  ] rabbitmq_peer_discovery_common    3.8.1
[  ] rabbitmq_peer_discovery_consul    3.8.1
[  ] rabbitmq_peer_discovery_etcd      3.8.1
[  ] rabbitmq_peer_discovery_k8s       3.8.1
[  ] rabbitmq_prometheus               3.8.1
[  ] rabbitmq_random_exchange          3.8.1
[  ] rabbitmq_recent_history_exchange  3.8.1
[  ] rabbitmq_sharding                 3.8.1
[E*] rabbitmq_shovel                   3.8.1
[E*] rabbitmq_shovel_management        3.8.1
[  ] rabbitmq_stomp                    3.8.1
[  ] rabbitmq_top                      3.8.1
[  ] rabbitmq_tracing                  3.8.1
[  ] rabbitmq_trust_store              3.8.1
[e*] rabbitmq_web_dispatch             3.8.1
[  ] rabbitmq_web_mqtt                 3.8.1
[  ] rabbitmq_web_mqtt_examples        3.8.1
[  ] rabbitmq_web_stomp                3.8.1
[  ] rabbitmq_web_stomp_examples       3.8.1
```

* Operating system, version, and patch level: Ubuntu 18.04, fully up to date.
</details>

We are running a setup with one local and one global rabbitmq server. The global one gets messages from a bunch of Web-Services, while the local one is connected to local on-location services.
To shovel messages between them, the local instance, which is the one described above, has a bunch of static shovels configured.
This works great, until at some seemingly random point in time, the shovels break. See the paste above for the crash log.
The shovels are all configured with a reconnect_delay of 5.0, but it seems like the max limit is reached _immediately_, and no reconnect attempt is ever made.
What's worse is that both the WebUI and `rabbitmqctl shovel_status` keep showing all the shovels as running, even though they clearly are not.

**Comment by michaelklishin** [comment]
I'm afraid we cannot suggest much without a set of steps to reproduce, which must include RabbitMQ version used, logs and, if possible, a Docker Compose or comparable setup that would let us reproduce this scenario reasonably reliably (Toxiproxy or iptables rules can be used to disrupt connections). Please help others help you.

**Comment by BtbN** [comment]
I provided logs and version info in the initial report, and put them into a details-block due to the length. It's behind the collapsible "System and Crash Info" thing.

**Comment by michaelklishin** [comment]
I tried reproducing using the following steps:

 * Set up a node with a `shovel:shovel` user on host B with sufficient permissions to access management UI so that I can inspect connections
 * Set up a node with Shovel enabled on node A
 * Set up a Toxiproxy node on node A that listens on `A:5673` and upstream at `B:5672`

The toxics on the proxy look like this:

```
Name			Listen		Upstream		Enabled		Toxics
======================================================================================
shovels			127.0.0.1:5673	mercurio.local:5672	enabled		1

Hint: inspect toxics with `toxiproxy-cli inspect <proxyName>`


Name: shovels	Listen: 127.0.0.1:5673	Upstream: mercurio.local:5672
======================================================================
Upstream toxics:
Proxy has no Upstream toxics enabled.

Downstream toxics:
timeout_downstream:	type=timeout	stream=downstream	toxicity=1.00	attributes=[	timeout=7000	]
```

This is the closest thing I can think of to trigger a `socket_closed_unexpectedly`. The fact
that I used dynamic Shovels should be irrelevant, the same goes for plain TCP connections
vs. TLS since socket errors in case of connection interruption described in this case would be identical.

Toxiproxy logs report numerous successful connections and reconnections:

```
INFO[0576] Accepted client                               client=127.0.0.1:56063 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0583] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56063: use of closed network connection name=shovels
WARN[0583] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56064->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0588] Accepted client                               client=127.0.0.1:56067 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0595] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56067: use of closed network connection name=shovels
WARN[0595] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56068->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0600] Accepted client                               client=127.0.0.1:56080 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0609] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56080: use of closed network connection name=shovels
WARN[0609] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56081->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0614] Accepted client                               client=127.0.0.1:56097 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0621] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56097: use of closed network connection name=shovels
WARN[0621] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56098->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0626] Accepted client                               client=127.0.0.1:56112 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0633] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56112: use of closed network connection name=shovels
WARN[0633] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56113->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0638] Accepted client                               client=127.0.0.1:56120 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0645] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56120: use of closed network connection name=shovels
WARN[0645] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56121->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0650] Accepted client                               client=127.0.0.1:56126 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0657] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56126: use of closed network connection name=shovels
WARN[0657] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56127->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0662] Accepted client                               client=127.0.0.1:56134 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0670] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56134: use of closed network connection name=shovels
WARN[0670] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56135->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0675] Accepted client                               client=127.0.0.1:56142 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0682] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56142: use of closed network connection name=shovels
WARN[0682] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56143->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0687] Accepted client                               client=127.0.0.1:56146 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0695] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56146: use of closed network connection name=shovels
WARN[0695] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56147->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0700] Accepted client                               client=127.0.0.1:56151 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0707] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56151: use of closed network connection name=shovels
WARN[0707] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56152->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0712] Accepted client                               client=127.0.0.1:56155 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0719] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56155: use of closed network connection name=shovels
WARN[0719] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56156->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0724] Accepted client                               client=127.0.0.1:56163 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0732] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56163: use of closed network connection name=shovels
WARN[0732] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56164->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0737] Accepted client                               client=127.0.0.1:56171 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0744] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56171: use of closed network connection name=shovels
WARN[0744] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56172->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0749] Accepted client                               client=127.0.0.1:56180 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0757] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56180: use of closed network connection name=shovels
WARN[0757] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56181->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0762] Accepted client                               client=127.0.0.1:56186 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0769] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56186: use of closed network connection name=shovels
WARN[0769] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56187->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0774] Accepted client                               client=127.0.0.1:56191 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0781] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56191: use of closed network connection name=shovels
WARN[0781] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56192->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0786] Accepted client                               client=127.0.0.1:56199 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0794] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56199: use of closed network connection name=shovels
WARN[0794] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56200->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0799] Accepted client                               client=127.0.0.1:56205 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0807] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56205: use of closed network connection name=shovels
WARN[0807] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56206->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0812] Accepted client                               client=127.0.0.1:56210 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0820] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56210: use of closed network connection name=shovels
WARN[0820] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56211->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0825] Accepted client                               client=127.0.0.1:56219 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0833] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56219: use of closed network connection name=shovels
WARN[0833] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56220->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0838] Accepted client                               client=127.0.0.1:56226 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
WARN[0847] Source terminated                             bytes=8 err=read tcp 127.0.0.1:5673->127.0.0.1:56226: use of closed network connection name=shovels
WARN[0847] Source terminated                             bytes=500 err=read tcp 192.168.1.142:56227->192.168.1.178:5672: use of closed network connection name=shovels
INFO[0852] Accepted client                               client=127.0.0.1:56238 name=shovels proxy=127.0.0.1:5673 upstream=mercurio.local:5672
```

So Shovels do reconnect as expected.

`reached_max_restart_intensity` suggests enough Shovels terminated in a period of time (which I'd have to look up in the code) to trigger an "unrecoverable condition" scenario. How many Shovels do you have running at peak times? My only two guesses are that

 * There are, say, hundreds of Shovels that terminate all at the same time
 * During recovery they immediately run into a `socket_closed_unexpectedly` but according to the logs it happens on the same second (unhandled exception logger does not provide millisecond precision timestamps).

**Comment by michaelklishin** [comment]
I removed the pastebin link since the exception contained sensitive information.

The gist comes down to

```
2019-11-29 23:41:14 =CRASH REPORT====
  crasher:
    initial call: amqp_gen_connection:init/1
    pid: <0.885.0>
    registered_name: []
    exception exit: {socket_closed_unexpectedly,[{gen_server,handle_common_reply,8,[{file,"gen_server.erl"},{line,751}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]}
    ancestors: [<0.883.0>,amqp_sup,<0.292.0>]
    message_queue_len: 0
    messages: []
    links: [<0.883.0>,<0.829.0>]
    dictionary: [{gen_server_call_timeout,60000},{process_name,{amqp_gen_connection,<<"client {source}:44745 -> {destination}:5671">>}}]
    trap_exit: true
    status: running
    heap_size: 2586
    stack_size: 27
    reductions: 25027
  neighbours:

2019-11-29 23:41:14 =SUPERVISOR REPORT====
     Supervisor: {<0.877.0>,amqp_connection_sup}
     Context:    shutdown
     Reason:     reached_max_restart_intensity
     Offender:   [{pid,<0.879.0>},{name,connection},{mfargs,{amqp_gen_connection,start_link,[<0.878.0>,{amqp_params_network, …}]
```

The `reached_max_restart_intensity` part is a red herring since it's an individual connection
that reaches its "restart limit" which is [set to 0](https://github.com/rabbitmq/rabbitmq-erlang-client/blob/master/src/amqp_connection_sup.erl#L50). Individual connections in the Erlang client
do not recover, their "owner" is responsible for that and the logs contain no Shovel log entries,
only Erlang client ones.

I will try to reproduce with a static Shovel but besides a mass disconnect don't have any other hypotheses.

**Comment by michaelklishin** [comment]
Besides sharing how many Shovels there are, can you please try to find any references to `amqp_sup` in the log? Or `rabbit_shovel_worker_sup`? and share those.

Or just paste more content with hostname and username redacted out? The password value logged in `3.8.1` is a value encrypted with an ephemeral key and not your Shovel's actual password but feel free to edit those out, too.

**Comment by michaelklishin** [comment]
This line:

```
ancestors: [<0.788.0>,moderation_shovel2,rabbit_shovel_sup,<0.688.0>]
```

is not what we are looking for. It's obvious that a Shovel is the owner of those connections. We need actual reports from `rabbit_shovel_sup`, `rabbit_shovel_worker_sup `, and especially `amqp_sup`, if any.

**Comment by michaelklishin** [comment]
Adding one more toxic, a data "slicer", did not make any difference in about ten minutes.
Neither does switching the toxics to upstream (which is more correct semantically, although if my understanding of the `timeout` toxic behavior is correct, it should not be relevant):

```

Name: shovels	Listen: 127.0.0.1:5673	Upstream: mercurio.local:5672
======================================================================
Upstream toxics:
Proxy has no Upstream toxics enabled.

Downstream toxics:
timeout_downstream:	type=timeout	stream=upstream	toxicity=1.00	attributes=[	timeout=7000	]
slicer_downstream:	type=slicer	stream=upstream	toxicity=1.00	attributes=[	average_size=112	delay=0	size_variation=0	]
```

**Comment by BtbN** [comment]
That's the (hopefully entirely redacted) log of the relevant timeframe: http://paste.ubuntu.com/p/cHdjjvpn8t/

It's also not a whole lot of shovels:

```
Shovel status on node rabbit@Encode-1
destination     destination_protocol    last_changed    name    source  source_protocol source_queue    state   termination_reason      type
amqp:///esa_local2      amqp091 2019-11-30 15:28:01     cg_shovel_1to2  amqp:///esa_local1      amqp091         running         static
amqp:///esa_local1      amqp091 2019-11-30 15:28:01     cg_shovel_2to1  amqp:///esa_local2      amqp091         running         static
amqps://global-hostname.tld/esa_prod    amqp091 2019-11-30 15:28:01     cg_shovel2      amqp:///esa_local2      amqp091         running         static
amqp:///esa_local2      amqp091 2019-11-30 15:28:01     moderation_shovel2      amqps://global-hostname.tld/esa_prod    amqp091         running         static
amqp:///esa_local2      amqp091 2019-11-30 15:28:01     tracker_shovel2 amqps://global-hostname.tld/esa_prod    amqp091         running         static
amqps://global-hostname.tld/esa_prod    amqp091 2019-11-30 15:28:01     cg_shovel1      amqp:///esa_local1      amqp091         running         static
amqp:///esa_local1      amqp091 2019-11-30 15:28:01     moderation_shovel1      amqps://global-hostname.tld/esa_prod    amqp091         running         static
amqp:///esa_local1      amqp091 2019-11-30 15:28:01     tracker_shovel1 amqps://global-hostname.tld/esa_prod    amqp091         running         static
amqps://global-hostname.tld/esa_prod    amqp091 2019-11-30 15:28:01     bigbutton_shovel2       amqp:///esa_local2      amqp091         running         static
amqps://global-hostname.tld/esa_prod    amqp091 2019-11-30 15:28:01     bigbutton_shovel1       amqp:///esa_local1      amqp091         running         static
```

**Comment by michaelklishin** [comment]
So there's a handful of Shovels.

I repeated my test with 1111 static Shovels generated like so:

``` ruby
<%- n = 1111 %>
[
  {rabbitmq_shovel,
      [
        {shovels, [
          <% (1..n).to_a.each do |i| %>
          {shovel<%= i %>,
                      [
                        {source,
                          [
                            {protocol, amqp091},
                            {uris, ["amqp://guest:guest@localhost:5672"]},
                            {queue, <<"static-shovel-source">>},
                            {reconnect_delay, 5},

                            {declarations, [
                                {'queue.declare',
                                   [{queue, <<"static-shovel-source">>},
                                    {durable, true}]}
                               ]}
                          ]},
                        {destination,
                          [
                            {protocol, amqp091},
                            {uris, ["amqp://shovel:shovel@localhost:5673"]},
                            {queue, <<"static-shovel-destination">>},
                            {ack_mode, on_confirm},
                            {reconnect_delay, 5},

                            {declarations, [
                                {'queue.declare',
                                   [{queue, <<"static-shovel-destination">>},
                                    {durable, true}]}
                               ]}
                      ]}
                  ]}<%= if i < n then "," else "" end %>
          <% end %>
          ]}
  ]}
].
```

(this is ERB). The result looks like this:

``` erlang

[
  {rabbitmq_shovel,
      [
        {shovels, [
          
          {shovel1,
                      [
                        {source,
                          [
                            {protocol, amqp091},
                            {uris, ["amqp://guest:guest@localhost:5672"]},
                            {queue, <<"static-shovel-source">>},
                            {reconnect_delay, 5},

                            {declarations, [
                                {'queue.declare',
                                   [{queue, <<"static-shovel-source">>},
                                    {durable, true}]}
                               ]}
                          ]},
                        {destination,
                          [
                            {protocol, amqp091},
                            {uris, ["amqp://shovel:shovel@localhost:5673"]},
                            {queue, <<"static-shovel-destination">>},
                            {ack_mode, on_confirm},
                            {reconnect_delay, 5},

                            {declarations, [
                                {'queue.declare',
                                   [{queue, <<"static-shovel-destination">>},
                                    {durable, true}]}
                               ]}
                      ]}
                  ]},
          
          {shovel2,
                      [
                        {source,
                          [
                            {protocol, amqp091},
                            {uris, ["amqp://guest:guest@localhost:5672"]},
                            {queue, <<"static-shovel-source">>},
                            {reconnect_delay, 5},

                            {declarations, [
                                {'queue.declare',
                                   [{queue, <<"static-shovel-source">>},
                                    {durable, true}]}
                               ]}
                          ]},
                        {destination,
                          [
                            {protocol, amqp091},
                            {uris, ["amqp://shovel:shovel@localhost:5673"]},
                            {queue, <<"static-shovel-destination">>},
                            {ack_mode, on_confirm},
                            {reconnect_delay, 5},

                            {declarations, [
                                {'queue.declare',
                                   [{queue, <<"static-shovel-destination">>},
                                    {durable, true}]}
                               ]}
                      ]}
                  ]}

                  %% more of the same with different IDs

  ]}
].
```

The behavior is the same except that it takes longer for 1111 connections to be set up sequentially,
so I had to bump the timeout. As soon as the toxic is removed, connections recover and behave as expected.

You can try using dynamic Shovels which are easier to monitor and restart (there's an HTTP API endpoint at [`DELETE /shovels/vhost/{vhost}/{name}/restart`](https://github.com/rabbitmq/rabbitmq-shovel-management/commit/6e1358416f3fb54e0bc73c9a463be28384178132) and worst case is restarting the plugin, not the entire node).

**Comment by BtbN** [comment]
What I so far did to workaround the issue was a cronjob that periodically checks the output of rabitmqctl shovel_status for "terminated", and if it's the case, it runs `rabbitmqctl eval 'application:stop(rabbitmq_shovel), application:start(rabbitmq_shovel).'`

But since it seems like it can get into a state where the shovels are disconnected but still showing up, that script did not catch tonight's incident.

We opted for static shovels because their configuration can be version controlled alongside our other configuration. I suppose a script which sets up the dynamic ones can be as well, but seemed more hacky to me at the time.

I can also confirm that under normal circumstances, the reconnecting seems to work just fine in our setup. There are also no signs for any internet connectivity issues during the time it failed.
Once our current event is over, I will try messing with it some more, intentionally dropping packages and the like.

**Comment by michaelklishin** [comment]
I don't see anything "hacky" about using an API for managing Shovels but if it's not convenient for you, there's nothing wrong with using static Shovels, it's just a more limited approach (but yes, it is version controllable without any additional effort). 

I'm curious if factors such as temporary hostname resolution or routing failures can be at play. The former would yield a different error, though, `nxdomain`.

Any additional metrics you can track and share would help. I tried every Toxiproxy toxic effect and Shovels still recover, even when there's over 1K of them.

I'll file an issue that would help you in the short term, an easy way to restart all static Shovels in a single command.

**Comment by BtbN** [comment]
I checked the logs on the global server, which for all of the shovels logs the exact same thing, at the exact same timestamp:
```
2019-11-29 22:41:08.647 [warning] <0.23321.0> closing AMQP connection <0.23321.0> (x.x.x.x:44745 -> z.z.z.z:5671 - Shovel moderation_shovel2, vhost: 'esa_prod', user: 'shove
Nov 29 22:41:08 esa-mq-1 docker-compose[14916]: rabbitmq_1  | client unexpectedly closed TCP connection
```

Curiously, one single shovel out of all of them survived it. Only to suffer the same fate 12 hours later.

I looked at the code which creates the "client unexpectedly closed TCP connection" message, because to me it sounds like the client (or something else) actively closed it, rather than a connection issue causing a disconnect, but I was unable to confirm if that's the case, or the same message is logged in any case.
